### ISSUE TITLE
[REVIEW]: BisPy: Bisimulation in Python

### ISSUE THREAD

**Issue by whedon** [openjournals/joss-reviews#3519]
**Submitting author:** @fAndreuzzi (<a href="http://orcid.org/0000-0002-9508-7801">Francesco Andreuzzi</a>)
**Repository:** <a href="https://github.com/fAndreuzzi/BisPy" target ="_blank">https://github.com/fAndreuzzi/BisPy</a>
**Version:** v0.2.3
**Editor:** @mjsottile
**Reviewers:** @jonjoncardoso, @mschordan, @zoometh
**Archive:** <a href="https://doi.org/10.5281/zenodo.5532910" target="_blank">10.5281/zenodo.5532910</a>

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

## Status

[![status](https://joss.theoj.org/papers/9d9c3ca0715d482938b5a450525cefa0/status.svg)](https://joss.theoj.org/papers/9d9c3ca0715d482938b5a450525cefa0)

Status badge code:

```
HTML: <a href="https://joss.theoj.org/papers/9d9c3ca0715d482938b5a450525cefa0"><img src="https://joss.theoj.org/papers/9d9c3ca0715d482938b5a450525cefa0/status.svg"></a>
Markdown: [![status](https://joss.theoj.org/papers/9d9c3ca0715d482938b5a450525cefa0/status.svg)](https://joss.theoj.org/papers/9d9c3ca0715d482938b5a450525cefa0)
```
**Reviewers and authors**:

Please avoid lengthy details of difficulties in the review thread. Instead, please create a new issue in the <a href="https://github.com/fAndreuzzi/BisPy" target="_blank">target repository</a> and link to those issues (especially acceptance-blockers) by leaving comments in the review thread below. (For completists: if the target issue tracker is also on GitHub, linking the review thread in the issue or vice versa will create corresponding breadcrumb trails in the link target.)

## Reviewer instructions & questions

@zoometh & @mschordan & @jonjoncardoso, please carry out your review in this issue by updating the checklist below. If you cannot edit the checklist please:
1. Make sure you're logged in to your GitHub account
2. Be sure to accept the invite at this URL: https://github.com/openjournals/joss-reviews/invitations

The reviewer guidelines are available here: https://joss.readthedocs.io/en/latest/reviewer_guidelines.html. Any questions/concerns please let @mjsottile know.

✨ **Please start on your review when you are able, and be sure to complete your review in the next six weeks, at the very latest** ✨ 

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fAndreuzzi/BisPy">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@fAndreuzzi) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fAndreuzzi/BisPy">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@fAndreuzzi) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [x] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [x] **Installation:** Does installation proceed as outlined in the documentation?
- [x] **Functionality:** Have the functional claims of the software been confirmed?
- [x] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [x] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [x] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [x] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [x] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [x] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [x] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [x] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [x] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [x] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [x] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [x] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

### Conflict of interest

- [x] I confirm that I have read the [JOSS conflict of interest (COI) policy](https://github.com/openjournals/joss/blob/master/COI.md) and that: I have no COIs with reviewing this work or that any perceived COIs have been waived by JOSS for the purpose of this review.

### Code of Conduct

- [x] I confirm that I read and will adhere to the [JOSS code of conduct](https://joss.theoj.org/about#code_of_conduct).

### General checks

- [x] **Repository:** Is the source code for this software available at the <a target="_blank" href="https://github.com/fAndreuzzi/BisPy">repository url</a>?
- [x] **License:** Does the repository contain a plain-text LICENSE file with the contents of an [OSI approved](https://opensource.org/licenses/alphabetical) software license?
- [x] **Contribution and authorship:** Has the submitting author (@fAndreuzzi) made major contributions to the software? Does the full list of paper authors seem appropriate and complete?
- [ ] **Substantial scholarly effort:** Does this submission meet the scope eligibility described in the [JOSS guidelines](https://joss.readthedocs.io/en/latest/submitting.html#substantial-scholarly-effort)

### Functionality

- [ ] **Installation:** Does installation proceed as outlined in the documentation?
- [ ] **Functionality:** Have the functional claims of the software been confirmed?
- [ ] **Performance:** If there are any performance claims of the software, have they been confirmed? (If there are no claims, please check off this item.)

### Documentation

- [ ] **A statement of need:** Do the authors clearly state what problems the software is designed to solve and who the target audience is?
- [ ] **Installation instructions:** Is there a clearly-stated list of dependencies? Ideally these should be handled with an automated package management solution.
- [ ] **Example usage:** Do the authors include examples of how to use the software (ideally to solve real-world analysis problems).
- [ ] **Functionality documentation:** Is the core functionality of the software documented to a satisfactory level (e.g., API method documentation)?
- [ ] **Automated tests:** Are there automated tests or manual steps described so that the functionality of the software can be verified?
- [ ] **Community guidelines:** Are there clear guidelines for third parties wishing to 1) Contribute to the software 2) Report issues or problems with the software 3) Seek support

### Software paper

- [ ] **Summary:** Has a clear description of the high-level functionality and purpose of the software for a diverse, non-specialist audience been provided?
- [ ] **A statement of need:** Does the paper have a section titled 'Statement of Need' that clearly states what problems the software is designed to solve and who the target audience is?
- [ ] **State of the field:** Do the authors describe how this software compares to other commonly-used packages?
- [ ] **Quality of writing:** Is the paper well written (i.e., it does not require editing for structure, language, or writing quality)?
- [ ] **References:** Is the list of references complete, and is everything cited appropriately that should be cited (e.g., papers, datasets, software)? Do references in the text use the proper [citation syntax]( https://rmarkdown.rstudio.com/authoring_bibliographies_and_citations.html#citation_syntax)?

**Comment by whedon** [comment]
Hello human, I'm @whedon, a robot that can help you with some common editorial tasks. @zoometh, @mschordan, @jonjoncardoso it looks like you're currently assigned to review this paper :tada:.

**:warning: JOSS reduced service mode :warning:**

Due to the challenges of the COVID-19 pandemic, JOSS is currently operating in a "reduced service mode". You can read more about what that means in [our blog post](https://blog.joss.theoj.org/2020/05/reopening-joss).

**:star: Important :star:**

If you haven't already, you should seriously consider unsubscribing from GitHub notifications for this (https://github.com/openjournals/joss-reviews) repository. As a reviewer, you're probably currently watching this repository which means for GitHub's default behaviour you will receive notifications (emails) for all reviews 😿

To fix this do the following two things:

1. Set yourself as 'Not watching' https://github.com/openjournals/joss-reviews:

![watching](https://cloud.githubusercontent.com/assets/4483/20250593/64d7ce48-a9de-11e6-9225-d3dfb3e48e68.png)

2. You may also like to change your default settings for this watching repositories in your GitHub profile here: https://github.com/settings/notifications

![notifications](https://cloud.githubusercontent.com/assets/4483/20250604/79eb3144-a9de-11e6-9e51-0fe940966689.png)


For a list of things I can do to help you, just type:

```
@whedon commands
```

For example, to regenerate the paper pdf after making changes in the paper's md or bib files, type:

```
@whedon generate pdf
```

**Comment by whedon** [comment]
Wordcount for `paper.md` is 988

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/3-540-44585-4_8 is OK
- 10.1023/A:1027328830731 is OK
- 10.1137/0201010 is OK
- 10.1016/B978-0-12-417750-5.50022-1 is OK
- 10.1137/0216062 is OK
- 10.1007/978-3-540-77050-3_17 is OK
- 10.1016/0898-1221(81)90008-0 is OK
- 10.1016/0890-5401(90)90025-D is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
```
Software report (experimental):

github.com/AlDanial/cloc v 1.88  T=0.08 s (719.2 files/s, 94295.7 lines/s)
-------------------------------------------------------------------------------
Language                     files          blank        comment           code
-------------------------------------------------------------------------------
Python                          37           1236           1297           4136
Markdown                         2             89              0            289
reStructuredText                14            120            157            199
TeX                              1              7              0             82
YAML                             2              6              6             48
DOS Batch                        1              8              1             26
make                             1              4              7              9
TOML                             1              1              0              8
-------------------------------------------------------------------------------
SUM:                            59           1471           1468           4797
-------------------------------------------------------------------------------


Statistical information for the repository 'd5f6644ece31ade8fd9a9522' was
gathered on 2021/07/21.
The following historical commit information, by author, was found:

Author                     Commits    Insertions      Deletions    % of changes
Francesco Andreuzzi              9           597            331            1.76
SV-97                            3           880            453            2.52
francescoandreuzzi             399         28541          22022           95.72

Below are the number of rows from each author that have survived and are still
intact in the current revision:

Author                     Rows      Stability          Age       % in comments
Francesco Andreuzzi        6092         1020.4          3.6               11.46
SV-97                       601           68.3          0.1                0.00
```

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: :point_left:

**Comment by mjsottile** [comment]
@whedon list reviewers

**Comment by whedon** [comment]
Here's the current list of reviewers: https://bit.ly/joss-reviewers

**Comment by mjsottile** [comment]
@whedon assign @zoometh as reviewer

**Comment by whedon** [comment]
OK, @zoometh is now a reviewer

**Comment by mjsottile** [comment]
@whedon add @jonjoncardoso as reviewer

**Comment by whedon** [comment]
OK, @jonjoncardoso is now a reviewer

**Comment by mjsottile** [comment]
@whedon add @zoometh as reviewer

**Comment by whedon** [comment]
OK, @zoometh is now a reviewer

**Comment by mjsottile** [comment]
@whedon add @mschordan as reviewer

**Comment by whedon** [comment]
OK, @mschordan is now a reviewer

**Comment by mjsottile** [comment]
:wave: @arfon : for some reason whedon isn’t adding @zoometh as an assigned reviewer for this submission even though they were added.  Do you know why?

**Comment by danielskatz** [comment]
@mjsottile - please use @openjournals/dev for technical problems, and `@openjournals/joss-eics` for the editors in chief for other things, in case one person, for example @arfon, is on vacation.  And I'll check on this now

**Comment by danielskatz** [comment]
@whedon remove @zoometh as reviewer

**Comment by whedon** [comment]
OK, @zoometh is no longer a reviewer

**Comment by danielskatz** [comment]
@whedon add @zoometh as reviewer

**Comment by whedon** [comment]
OK, @zoometh is now a reviewer

**Comment by danielskatz** [comment]
Ok, this is now fine, as seen in the first comment in this issue, where it now says `Reviewers: @jonjoncardoso, @mschordan, @zoometh`  When I started, zoometh was listed twice, since you had added them twice, so I removed them then readded them so they would only be listed once.  The whedon command also shows this correctly in its response immediately above, but this is less important than what the first comment shows.  Now you have to manually edit the first comment and add a checklist for zoometh, as whedon only generates these checklists once, when the review issue is created, based on the reviewers that were assigned at the time (in the pre-review issue). If you need help with this, let `@openjournals/joss-eics` know.

**Comment by mjsottile** [comment]
@danielskatz thanks.  When I looked at it last night the reviewer did show up in the issue body and had a checklist, but was missing from the list of assignees for the GitHub issue itself.  They appear to be listed as an assignee now.

**Comment by zoometh** [comment]
@fAndreuzzi I'll suggest reordering the parts of your paper & to modify some:

1. **Summary**: make it shorter, currently it is too specific (ie not enough general)
2. **Statement of need**: has to be placed _before_ the **Examples**
3. ~ **BisPy framework**: present your Python package here
4. **Examples**
5. ...

I've also noted:
- There is a missing [community guideline](https://joss.readthedocs.io/en/latest/review_checklist.html#documentation) for contributions' Code of conduct
- Few typo errors:
  -  in [README](https://github.com/fAndreuzzi/BisPy#dependencies-and-installation): "_requires requires_"
  - I'm not sure to understand this sentence "_Again, the two nodes are almost indistinguishable: we may switch `b` and `b`, and unless `a` resents the node `c` for some reason nobody would notice_" in the [docs/](https://github.com/fAndreuzzi/BisPy/blob/master/docs/source/index.rst), maybe a typo error

**Comment by fAndreuzzi** [comment]
Hi @zoometh, thank you for your feedback.

I worked a little bit on the README (fixed the typo you mentioned, and added a [new](https://github.com/fAndreuzzi/BisPy/blob/master/README.md#reporting-a-bug) section to explain in detail how to report a bug. In [this](https://github.com/fAndreuzzi/BisPy#how-to-contribute) section I explain how to submit a pull request, is this enough for you?

I'm going to work on the paper to apply your suggestions, I'll update you here when I'm satisfied with the changes.

**Comment by fAndreuzzi** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: :point_left:

**Comment by fAndreuzzi** [comment]
Following the suggestions by @zoometh I moved up the "Statement of need", and shrinked the dimension of the summary. Let me know what you think, and if you have any other suggestion.

**Comment by whedon** [comment]
:wave: @jonjoncardoso, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @mschordan, please update us on how your review is going (this is an automated reminder).

**Comment by whedon** [comment]
:wave: @zoometh, please update us on how your review is going (this is an automated reminder).

**Comment by fAndreuzzi** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: :point_left:

**Comment by fAndreuzzi** [comment]
@zoometh In the last version I added the section **BisPy** and slightly modified the **Statement of need** accordingly. What do you think?

**Comment by zoometh** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/3-540-44585-4_8 is OK
- 10.1023/A:1027328830731 is OK
- 10.1137/0201010 is OK
- 10.1016/B978-0-12-417750-5.50022-1 is OK
- 10.1137/0216062 is OK
- 10.1007/978-3-540-77050-3_17 is OK
- 10.1016/0898-1221(81)90008-0 is OK
- 10.1016/0890-5401(90)90025-D is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by zoometh** [comment]
> 
> 
> @zoometh In the last version I added the section **BisPy** and slightly modified the **Statement of need** accordingly. What do you think?

@fAndreuzzi  All right, that's fine for me. Great job.

**Comment by jonjoncardoso** [comment]
Hi! I haven't been able to properly start reviewing of the paper and code yet but I should do that soon in the next few days :)

**Comment by fAndreuzzi** [comment]
Hi @jonjoncardoso and @mschordan, since it has been a while since the last activity on this issue I would like to know if you encountered issues during the installation or usage of the package, or while gathering the information needed for the review. Please, let me know if I can help in some way!

**Comment by mschordan** [comment]
@fAndreuzzi The README.md file states that:
>>> compute_maximum_bisimulation(graph, initial_partition=[(0,7,10), (1,2,3,4,5,6,8,9,11,12,13,14)])
should produce: [(7, 10), (8, 9, 11, 12, 13, 14), (3, 4, 5, 6),  (1, 2), (0,)]
but I get: [(7, 10), (5, 6), (8, 9, 11, 12, 13, 14), (3, 4), (2,), (1,), (0,)]
all preceding examples give the expected result.

**Comment by fAndreuzzi** [comment]
Thank you very much @mschordan, the right output is indeed the one given by **BisPy**, I probably confused the outputs while doing copy-paste from the terminal to the browser. I fixed `README.md` in my last commit.

**Comment by jonjoncardoso** [comment]
@fAndreuzzi I haven't gone through the code yet but some general comments so far:

- I know you have pointed to reference elsewhere but I feel like some references are also needed to support this sentence about the problem complexity: _"[...] several algorithms for the computation of maximum bisimulation have been studied throughout the years, and it is now widely known that the problem has an O(|E| log |V |) algorithmic solution"_ (Line 15). 
- I suggest that recent changes you made to the summary and README should also be replicated to your readthedocs page. The description of bisimulation problem is clearer on the paper manuscript version compared to the page, for instance.

**Comment by fAndreuzzi** [comment]
> some references are also needed to support this sentence about the problem complexity

Sure, I'm still not completely familiar with the rule "cite whenever you say something which is not obvious, I fixed the problem in one of the latest commits.

> The description of bisimulation problem is clearer on the paper manuscript version

Could you please point me the parts where I explained the bisimulation better in the paper rather than in the documentation? In my opinion the explanation in the paper is very limited. Anyway I improved a little bit the explanation on [ReadTheDocs](https://bispy-bisimulation-in-python.readthedocs.io/en/latest/) (cleared some sentences and added a new paragraph about maximum bisimulation). Kindly let me know what you think.

Thank you

**Comment by mjsottile** [comment]
:wave: @jonjoncardoso @mschordan Let me know if you need anything to continue making progress on your reviews.  Don't forget to update your checklists at the top of this review issue so I can track review progress.

**Comment by mschordan** [comment]
@fAndreuzzi 
1) You write: "We found some sparse implementations of the most famous algorithm for the computation of the maximum bisimulation (Paige-Tarjan), however the source code lacked documentation and test cases, and did not seem to be intended for distribution."

This may sound strange to a reader. How did you find it, when it was not intended for distribution? Can you provide a reference? If not, it's not clear why this is mentioned in the first place.

2) You write: "Other algorithms to compute the maximum bisimulation are available, but for what we could see they are slight variations of one of those mentioned above, or tailored on particular cases."

Which algorithms are you referring to? What are "slight variations"? Any algorithms that are referred to should be cited properly.

3) You write: "Our implementations have been tested and documented deeply;" - That's indeed an important aspect as you mention also that "our package may be useful to assess the performance". Some more information to make this point could be interesting to a potential user.
What was the biggest input that you tested the algorithm on? Can you provide some runtime measurement?

**Comment by fAndreuzzi** [comment]
Hi @mschordan, sorry for the delay but I was a little bit busy this week.

> How did you find it, when it was not intended for distribution?

I found code snippets on GitHub, but as I mention in the paper they lacked documentation, test cases, and a brief examination of the code made me think that time complexity was not taken into account at all. When I say "_not intended for distribution_" I mean that those snippets are not meant to be included in someone else's code or used for experiments whose intent is something more than "_what is the maximum bisimulation of this graph?_". Maybe I should change "_distribution_" to something else in order to express the meaning in a better way?

> Which algorithms are you referring to? What are "slight variations"? Any algorithms that are referred to should be cited properly.

If this sentence causes problems I think I can remove it from the paper. I do not think that it is interesting or meaningful to cite algorithms which are not those that I implemented. 

> What was the biggest input that you tested the algorithm on? Can you provide some runtime measurement?

Sure, do you think I should rather include runtime measurement in `readme.md` or in the paper?

**Comment by mschordan** [comment]
@fAndreuzzi 
> > How did you find it, when it was not intended for distribution?
> 
> I found code snippets on GitHub, but as I mentioned in the paper they lacked documentation, test cases, and a brief examination of the code made me think that time complexity was not taken into account at all. When I say "_not intended for distribution_" I mean that those snippets are not meant to be included in someone else's code or used for experiments whose intent is something more than "_what is the maximum bisimulation of this graph?_". Maybe I should change "_distribution_" to something else in order to express the meaning in a better way?

If you want to mention those, then there should also be a link to the respective GitHub repo, otherwise it's not clear what you are referring to. Based on whether you want to provide a reference/link you can decide whether you want to mention it.

> 
> > Which algorithms are you referring to? What are "slight variations"? Any algorithms that are referred to should be cited properly.
> 
> If this sentence causes problems I think I can remove it from the paper. I do not think that it is interesting or meaningful to cite algorithms which are not those that I implemented.

Indeed.

> > What was the biggest input that you tested the algorithm on? Can you provide some runtime measurement?
> 
> Sure, do you think I should rather include runtime measurement in `readme.md` or in the paper?

Where you mention that "our package may be useful to assess the performance", there it would be good to also provide some numbers on the runtime for a given larger input (and which machine it was run on). This would be good to be mentioned in the paper. Also when you mention "tested deeply", it's better to provide some actual info, e.g. how many tests you provide and what your biggest test input was, how the tests were selected, etc.

**Comment by fAndreuzzi** [comment]
@whedon generate pdf

**Comment by whedon** [comment]
:point_right::page_facing_up: [Download article proof](https://raw.githubusercontent.com/openjournals/joss-papers/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: [View article proof on GitHub](https://github.com/openjournals/joss-papers/blob/joss.03519/joss.03519/10.21105.joss.03519.pdf) :page_facing_up: :point_left:

**Comment by fAndreuzzi** [comment]
Hi @mschordan, thank you for your suggestions, I hopefully fixed/added everything you mentioned in your last comment. What do you think?

**Comment by mschordan** [comment]
@fAndreuzzi All changes are fine, looks all OK now.

**Comment by fAndreuzzi** [comment]
Hi @jonjoncardoso, I would like to notify that I think I fixed the problems your mentioned in your last comment, please let me know if there is something more you think should be improved.

**Comment by mjsottile** [comment]
:wave: @jonjoncardoso Let me know if you need anything to make progress on your review - I noticed it's been a month or so since your last review activity.

**Comment by fAndreuzzi** [comment]
Hi @mjsottile, I would like to ask if you could consider the possibility of considering the review over even though there are several unticked checkboxes. I see this happened some times on JOSS, and that the average number of reviewers per paper is usually 2.

I'm proposing this solution since we have no information about a date in which the last reviewer might have some free time to complete the review.

**Comment by mjsottile** [comment]
Hi @fAndreuzzi : that is my plan.  When a reviewer disappears with a partial review, but we have other completed reviews, I usually give them a few days after my last ping to see if they return.  Otherwise I carry on without it.  I believe we have enough with the completed reviews - I just need to give it a day or so until I have time to do my own final pass over the completed reviews and the paper itself.  I expect to have time to get to that by the end of the weekend or early in the week.

**Comment by fAndreuzzi** [comment]
@mjsottile Okay, this sounds like a good plan to me. Thank you very much for the fast reply!

**Comment by jonjoncardoso** [comment]
Hi @fAndreuzzi and @mjsottile , so sorry for the delay. Arrangements for the beginning of academic term was to blame here. 

It is ok from me as well, I think the paper and package is ready.

**Comment by mjsottile** [comment]
Excellent.  @fAndreuzzi - I have done a pass over the paper and have found a couple of minor grammatical mistakes to correct:

- line 43-44 : "moreover we splitted the algorithms in smaller functions" should read "moreover we split the algorithms into smaller functions".
- line 86 : "taken by step 4. on a sample of" -- remove the period after the digit 4.
- line 88 : "interests" should be "interest"
- line 90 : "to remark this improvement".  I believe the word "remark" should be replaced with the word "illustrate".
- line 92 : "operative system" should be "operating system"

Once you've finished making these corrections, I can proceed to the final steps.

The final step in the process before I hand it off to the EIC team is to make a tagged release and archive. JOSS needs the software to be archived in an archival repository, such as Zenodo, figshare, or an institutional repository. Once this is done, that repository should provide a DOI to the archive. Please post the DOI that you receive in this thread so I can add it to the submission. Make sure that the version of the software on this review issue matches the version of the software that you archive - let me know if I need to update it.

Once you have completed this, I'll run the final steps and hand the submission off to get it finished up.

**Comment by fAndreuzzi** [comment]
@whedon generate pdf

**Comment by fAndreuzzi** [comment]
@mjsottile Thank you for your review, I uploaded my code to Zenodo (version **0.2.3**, which is different than the one which I read in the header of this issue).

DOI: 10.5281/zenodo.5532910

**Comment by mjsottile** [comment]
@whedon set v0.2.3 as version

**Comment by whedon** [comment]
OK. v0.2.3 is the version.

**Comment by mjsottile** [comment]
@whedon set 10.5281/zenodo.5532910 as archive

**Comment by whedon** [comment]
OK. <a href="https://doi.org/10.5281/zenodo.5532910" target="_blank">10.5281/zenodo.5532910</a> is the archive.

**Comment by mjsottile** [comment]
@whedon check references

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/3-540-44585-4_8 is OK
- 10.1023/A:1027328830731 is OK
- 10.1137/0201010 is OK
- 10.1016/B978-0-12-417750-5.50022-1 is OK
- 10.1137/0216062 is OK
- 10.1007/978-3-540-77050-3_17 is OK
- 10.1016/0898-1221(81)90008-0 is OK
- 10.1016/0890-5401(90)90025-D is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by mjsottile** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Attempting dry run of processing paper acceptance...
```

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/3-540-44585-4_8 is OK
- 10.1023/A:1027328830731 is OK
- 10.1137/0201010 is OK
- 10.1016/B978-0-12-417750-5.50022-1 is OK
- 10.1137/0216062 is OK
- 10.1007/978-3-540-77050-3_17 is OK
- 10.1016/0898-1221(81)90008-0 is OK
- 10.1016/0890-5401(90)90025-D is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2615

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2615, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@fAndreuzzi - I've suggested some small changes in the paper and bib in https://github.com/fAndreuzzi/BisPy/pull/36 - please merge this, or let me know what you disagree with, then we can proceed to acceptance and publishing.

**Comment by fAndreuzzi** [comment]
@danielskatz Thank you very much for your changes, I merged your PR.

**Comment by danielskatz** [comment]
@whedon recommend-accept

**Comment by whedon** [comment]
```
Reference check summary (note 'MISSING' DOIs are suggestions that need verification):

OK DOIs

- 10.1007/3-540-44585-4_8 is OK
- 10.1023/A:1027328830731 is OK
- 10.1137/0201010 is OK
- 10.1016/B978-0-12-417750-5.50022-1 is OK
- 10.1137/0216062 is OK
- 10.1007/978-3-540-77050-3_17 is OK
- 10.1016/0898-1221(81)90008-0 is OK
- 10.1016/0890-5401(90)90025-D is OK

MISSING DOIs

- None

INVALID DOIs

- None
```

**Comment by whedon** [comment]
:wave: @openjournals/joss-eics, this paper is ready to be accepted and published.

 Check final proof :point_right: https://github.com/openjournals/joss-papers/pull/2623

If the paper PDF and Crossref deposit XML look good in https://github.com/openjournals/joss-papers/pull/2623, then you can now move forward with accepting the submission by compiling again with the flag `deposit=true` e.g.
 ```
@whedon accept deposit=true
```

**Comment by danielskatz** [comment]
@whedon accept deposit=true

**Comment by whedon** [comment]
```
Doing it live! Attempting automated processing of paper acceptance...
```

**Comment by whedon** [comment]
🐦🐦🐦 👉 [Tweet for this paper](https://twitter.com/JOSS_TheOJ/status/1442843765044023299) 👈 🐦🐦🐦

**Comment by whedon** [comment]
🚨🚨🚨 **THIS IS NOT A DRILL, YOU HAVE JUST ACCEPTED A PAPER INTO JOSS!** 🚨🚨🚨

 Here's what you must now do:

0. Check final PDF and Crossref metadata that was deposited :point_right: https://github.com/openjournals/joss-papers/pull/2624
1. Wait a couple of minutes, then verify that the paper DOI resolves [https://doi.org/10.21105/joss.03519](https://doi.org/10.21105/joss.03519)
2. If everything looks good, then close this review issue.
3. Party like you just published a paper! 🎉🌈🦄💃👻🤘

 Any issues? Notify your editorial technical team...

**Comment by fAndreuzzi** [comment]
Thank you very much to all the reviewers and editors!

**Comment by danielskatz** [comment]
Congratulations to @fAndreuzzi (Francesco Andreuzzi)!!

And thanks to @mschordan, @zoometh for reviewing, and @mjsottile for editing!
We couldn't do this without you!

**Comment by whedon** [comment]
:tada::tada::tada: Congratulations on your paper acceptance! :tada::tada::tada:

If you would like to include a link to your paper from your README use the following code snippets:

```
Markdown:
[![DOI](https://joss.theoj.org/papers/10.21105/joss.03519/status.svg)](https://doi.org/10.21105/joss.03519)

HTML:
<a style="border-width:0" href="https://doi.org/10.21105/joss.03519">
  <img src="https://joss.theoj.org/papers/10.21105/joss.03519/status.svg" alt="DOI badge" >
</a>

reStructuredText:
.. image:: https://joss.theoj.org/papers/10.21105/joss.03519/status.svg
   :target: https://doi.org/10.21105/joss.03519
```
This is how it will look in your documentation:

[![DOI](https://joss.theoj.org/papers/10.21105/joss.03519/status.svg)](https://doi.org/10.21105/joss.03519)

**We need your help!**

Journal of Open Source Software is a community-run journal and relies upon volunteer effort. If you'd like to support us please consider doing either one (or both) of the the following:

- Volunteering to review for us sometime in the future. You can add your name to the reviewer list here: https://joss.theoj.org/reviewer-signup.html
- Making a small donation to support our running costs here: https://numfocus.org/donate-to-joss
